packages/actor-query-operation-leftjoin-left-deep/lib/ActorQueryOperationLeftJoinLeftDeep.ts

```diff
@@ -0,0 +1,151 @@
+        };
+      },
+      pattern: (op: Algebra.Pattern, factory: Factory) => {
+        return {
+          recurse: false,
+          result: factory.createPattern(
+            ActorQueryOperationLeftJoinLeftDeep.materializeTerm(op.subject, bindings),
+            ActorQueryOperationLeftJoinLeftDeep.materializeTerm(op.predicate, bindings),
+            ActorQueryOperationLeftJoinLeftDeep.materializeTerm(op.object, bindings),
+            ActorQueryOperationLeftJoinLeftDeep.materializeTerm(op.graph, bindings),
+          ),
+        };
+      },
+    });
+  }
+
+  /**
+   * Materialize a term with the given binding.
+   *
+   * If the given term is a variable (or blank node)
+   * and that variable exist in the given bindings object,
+   * the value of that binding is returned.
+   * In all other cases, the term itself is returned.
+   *
+   * @param {RDF.Term} term A term.
+   * @param {Bindings} bindings A bindings object.
+   * @return {RDF.Term} The materialized term.
+   */
+  public static materializeTerm(term: RDF.Term, bindings: Bindings): RDF.Term {
+    if (term.termType === 'Variable') {
+      const value: RDF.Term = bindings.get(termToString(term));
+      if (value) {
+        return value;
+      }
+    }
+    return term;
+  }
+
+  public async testOperation(pattern: Algebra.LeftJoin, context: ActionContext): Promise<IActorTest> {
+    return true;
+  }
+
+  public async runOperation(pattern: Algebra.LeftJoin, context: ActionContext)
+    : Promise<IActorQueryOperationOutputBindings> {
+    // Initiate left and right operations
+    // Only the left stream will be used.
+    // The right stream is ignored and only its metadata and variables are used.
+    const left = ActorQueryOperation.getSafeBindings(await this.mediatorQueryOperation
+      .mediate({ operation: pattern.left, context }));
+    const right = ActorQueryOperation.getSafeBindings(await this.mediatorQueryOperation
+      .mediate({ operation: pattern.right, context }));
+
+    // Close the right stream, since we don't need that one
+    right.bindingsStream.close();
+
+    // If an expression was defined, wrap the right operation in a filter expression.
+    const rightOperation = pattern.expression
+      ? ActorQueryOperationLeftJoinLeftDeep.FACTORY.createFilter(pattern.right, pattern.expression)
+      : pattern.right;
+
+    // Create a left-deep stream with left and right.
+    const bindingsStream = ActorQueryOperationLeftJoinLeftDeep.createLeftDeepStream(left.bindingsStream, rightOperation,
+      async (operation: Algebra.Operation) => ActorQueryOperation.getSafeBindings(
+        await this.mediatorQueryOperation.mediate({ operation, context })).bindingsStream);
+
+    // Determine variables and metadata
+    const variables = ActorRdfJoin.joinVariables({ entries: [left, right] });
+    const metadata = () => Promise.all([left, right].map((entry) => entry.metadata()))
+      .then((metadatas) => metadatas.reduce((acc, val) => acc * val.totalItems, 1))
```

Comment:
I would have just written `metadatas[0].totalItems * metadatas[1].totalItems` but I guess this also works :D